### PR TITLE
feat(repository): created hasAndBelongsToMany following convention pa…

### PR DESCRIPTION
**Pull request by codejamninja**
I created some boilerplate for the hasAndBelongsToMany relation using the hasMany relation. Currently this contains the identical functionality as the hasMany relation.

See also #2308

## Checklist

- [x] `npm test` passes on your machine
- [ ] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

### CODE REVIEW

**Comment by TomerSalton** [comment]
hey @codejamninja ,
Since we are both assigned on issue #2043, I would like us to cooperate.
Let's avoid doing duplicated work and try to speed up the development process.

I'm currently working on the design and on the repository interface.
Can we agree to share our progress and knowledge in here?

**Comment by codejamninja** [comment]
@TomerSalton, Absolutely! Maybe we should sync up over a call sometime too.

**Comment by guitarhero1** [comment]
up

**Comment by dougal83** [comment]
> up

@guitarhero1 Hey, please don't make meaningless posts that contribute nothing to the conversation.  If you are attempting to "bump" the topic, please consider simply registering interest using the reaction feature.  All you've done here is add a pointless notification to everyone watching the thread(as have I).  This code will likely be reviewed as workloads allow, not on demand.

Apologies for the extra notification form this post.

File: packages/repository/src/repositories/legacy-juggler-bridge.ts [hunk 1]
```diff
@@ -210,6 +213,71 @@ export class DefaultCrudRepository<T extends Entity, ID>
     );
   }
 
+  /**
+   * @deprecated
```
Comment (bajtos) [inline packages/repository/src/repositories/legacy-juggler-bridge.ts]: Let's not add methods that are already deprecated at creation time. Please remove `_createHasAndBelongsToManyRepositoryFactoryFor` from the pull request.

File: packages/repository/src/relations/has-and-belongs-to-many/has-and-belongs-to-many.repository.ts [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  TargetRepository extends EntityCrudRepository<TargetEntity, TargetID>
+> implements HasAndBelongsToManyRepository<TargetEntity> {
+  /**
+   * Constructor of DefaultHasAndBelongsToManyEntityCrudRepository
+   * @param getTargetRepository the getter of the related target model repository instance
+   * @param constraint the key value pair representing foreign key name to constrain
+   * the target repository instance
+   */
+  constructor(
+    public getTargetRepository: Getter<TargetRepository>,
```
Comment (bajtos) [inline packages/repository/src/relations/has-and-belongs-to-many/has-and-belongs-to-many.repository.ts]: I am not sure if it's enough to provide target repository getter, don't we also need getter for the intermediate ("through") repository?

**Comment by raymondfeng** [comment]
@codejamninja Thank you so much for the efforts and patience. I'll find some time to review it tomorrow.

**Comment by codejamninja** [comment]
@raymondfeng no worries and thanks. If you wouldn't mind, review the `hasManyThrough` using `hasMany` first. This pull request cannot be finished until `hasManyThrough` is accepted.

#2359

**Comment by dhmlau** [comment]
We just switch the contribution method from CLA to DCO, making your contribution easier in the future. Please sign the commits with DCO by amending your commit messages with `-s` flag and push the changes again. If you're using the command line, you can:

```
git commit --amend -s
git push --force-with-lease
```

Please refer to [this docs page](https://loopback.io/doc/en/contrib/code-contrib.html#developer-certificate-of-origin-dco) for details. Thanks!

**Pull request closed without merging**